Content.Server/Nyanotrasen/StationEvents/Events/GlimmerShadeSpawnRule.cs

```diff
@@ -0,0 +1,53 @@
+using System.Linq;
+using Robust.Shared.Random;
+using Content.Server.GameTicking.Rules.Components;
+using Content.Shared.Atmos.Miasma;
+using Content.Server.StationEvents.Components;
+using Content.Shared.Psionics.Glimmer;
+
+namespace Content.Server.StationEvents.Events;
+
+internal sealed class GlimmerShadeSpawnRule : StationEventSystem<GlimmerShadeRuleComponent>
+{
+    [Dependency] private readonly IRobustRandom _robustRandom = default!;
+    [Dependency] private readonly GlimmerSystem _glimmerSystem = default!;
+
+    private static readonly string ShadePrototype = "MobShade";
+
+    protected override void Started(EntityUid uid, GlimmerShadeRuleComponent component, GameRuleComponent gameRule, GameRuleStartedEvent args)
+    {
+        base.Started(uid, component, gameRule, args);
+
+        var rottingPlaces = EntityManager.EntityQuery<RottingComponent, TransformComponent>();
+        var normalSpawnLocations = EntityManager.EntityQuery<VentCritterSpawnLocationComponent, TransformComponent>().ToList().ConvertAll(item => item.Item2.Coordinates);
+        var hiddenSpawnLocations = EntityManager.EntityQuery<MidRoundAntagSpawnLocationComponent, TransformComponent>().ToList().ConvertAll(item => item.Item2.Coordinates);
+
+        var spawnLocations = normalSpawnLocations.ToHashSet();
+        spawnLocations.UnionWith(hiddenSpawnLocations.ToHashSet());
+
+        if (spawnLocations.Count == 0)
+            return;
+
+        int guaranteedSpawns = _robustRandom.Next(1, (int) _glimmerSystem.GetGlimmerTier() + 1);
+
+        int i = 0;
+        while (i < guaranteedSpawns)
+        {
+            Spawn(ShadePrototype, _robustRandom.Pick(spawnLocations));
+            i++;
```

Comment:
Could replace `i` with `guaranteeedSpawns` as a decrementer. Could rename to `spawnsLeft` or some fitting name to clarify it's acting as one if you prefer that.